File: Discovery.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Windows.Devices.Bluetooth;
using Windows.Devices.Enumeration;
using Windows.Devices.HumanInterfaceDevice;

namespace Kontro
{
    public sealed class ControllerInfo
    {
        /// <summary>Endereco Bluetooth. Zero quando o controle so aparece por cabo.</summary>
        public ulong Address { get; set; }
        public string Name { get; set; }
        public DateTime LastSeen { get; set; }

        /// <summary>Interface HID do controle, usada para ler a carga sem Bluetooth.</summary>
        public string HidId { get; set; }

        /// <summary>Id do dispositivo no Windows, para consultar a carga que ele mantem.</summary>
        public string InstanceId { get; set; }

        /// <summary>
        /// Slot do XInput, quando o controle so existe por ali. Vale -1 para quem foi
        /// encontrado como dispositivo HID.
        /// </summary>
        public int XInputSlot { get; set; } = -1;

        /// <summary>
        /// Container do aparelho fisico. E o que amarra as varias interfaces do mesmo
        /// controle -- HID, XUSB, no do dispositivo -- e por isso serve para achar a
        /// carga mesmo quando ela nao esta no no que o app abriu.
        /// </summary>
        public string ContainerId { get; set; }

        /// <summary>
        /// Identidade estavel do controle. Com Bluetooth e o endereco; sem ele, o
        /// proprio id da interface, que ja carrega fabricante, produto e uma parte
        /// especifica daquela conexao.
        /// </summary>
        public string Key =>
            Address != 0 ? Address.ToString("x12")
            : XInputSlot >= 0 ? "xinput:" + XInputSlot
            : "hid:" + ChaveDoHid(HidId);

        private static string ChaveDoHid(string id)
        {
            if (string.IsNullOrEmpty(id)) return "desconhecido";
            var limpo = new System.Text.StringBuilder(id.Length);
            foreach (var c in id)
                if (char.IsLetterOrDigit(c)) limpo.Append(char.ToLowerInvariant(c));
            var texto = limpo.ToString();
            return texto.Length <= 40 ? texto : texto[^40..];
        }

        public string PrettyAddress => Address == 0
            ? null
            : string.Join(":", Enumerable.Range(0, 6)
                .Select(i => ((Address >> ((5 - i) * 8)) & 0xFF).ToString("X2")));
    }

    /// <summary>
    /// Descobre controles sem depender de marca ou modelo: pega os dispositivos HID cuja
    /// usage e gamepad ou joystick, extrai o endereco Bluetooth do id da interface e cruza
    /// com os dispositivos Bluetooth pareados. O nome exibido vem do proprio sistema.
    /// </summary>
    internal static class Discovery
    {
        private const ushort UsagePageGenericDesktop = 0x01;
        private const ushort UsageJoystick = 0x04;
        private const ushort UsageGamepad = 0x05;
        private const ushort UsageMultiAxis = 0x08;

        // um bloco de 12 hex delimitado por & ou _ dentro do id da interface HID
        private static readonly Regex AddressPattern =
            new(@"[&_]([0-9A-Fa-f]{12})[&_]", RegexOptions.Compiled);

        internal static async Task<List<ControllerInfo>> DiscoverAsync()
        {
            // Sem HID ainda pode haver controle: o XUSB e um mundo a parte, conferido
            // mais abaixo. Desistir aqui era justamente o que escondia esses controles.
            var encontrados = await GamepadsHidAsync();

            // Nomes bons vem do Bluetooth: o HID costuma devolver rotulos genericos
            // como "Controlador de jogo compativel com HID".
            var porEndereco = new Dictionary<ulong, string>();
            try
            {
                var pareados = await DeviceInformation.FindAllAsync(
                    BluetoothLEDevice.GetDeviceSelectorFromPairingState(true));
                foreach (var info in pareados)
                {
                    BluetoothLEDevice dev = null;
                    try { dev = await BluetoothLEDevice.FromIdAsync(info.Id); }
                    catch { }
                    if (dev == null) continue;
                    if (!string.IsNullOrWhiteSpace(dev.Name)) porEndereco[dev.BluetoothAddress] = dev.Name;
                    dev.Dispose();
                }
            }
            catch { }

            var resultado = new List<ControllerInfo>();
            var containersHid = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (var g in encontrados)
            {
                // sem endereco significa cabo ou dongle de radio: continua sendo um
                // controle, so nao tem leitura por Bluetooth
                string nome = g.Endereco != 0 && porEndereco.TryGetValue(g.Endereco, out var bom)
                    ? bom
                    : (string.IsNullOrWhiteSpace(g.Nome) ? "Controle" : g.Nome);

                resultado.Add(new ControllerInfo
                {
                    Address = g.Endereco,
                    Name = nome,
                    HidId = g.IdHid,
                    InstanceId = g.IdInstancia,
                    ContainerId = g.Container,
                    LastSeen = DateTime.Now
                });
                containersHid.Add(g.Container);
            }

            await AcrescentarSomenteXInputAsync(resultado, containersHid);
            return resultado;
        }

        /// <summary>Interface que o Windows publica para todo controle atendido pelo XUSB.</summary>
        private const string GuidXusb = "{EC87F1E3-C13B-4100-B5F7-8B84D54260CB}";
        private const string PropContainer = "System.Devices.ContainerId";
        private const string PropInstancia = "System.Devices.DeviceInstanceId";

        /// <summary>
        /// Acrescenta controles que existem apenas para o XInput.
        ///
        /// Quem usa o driver do Xbox 360 -- o caso dos dongles que emulam esse controle,
        /// e do que o Windows lista como "Xbox 360 for Windows" -- nao e dispositivo HID.
        /// A busca por HID passa direto por ele, e sem isto esse controle simplesmente
        /// nao existe para o app.
        ///
        /// A comparacao e por container, que e o aparelho fisico: um controle atendido
        /// por HID e por XUSB publica as duas interfaces sob o mesmo container. Container
        /// de XUSB que nao apareceu no HID e, por definicao, controle que so o XInput
        /// enxerga -- criterio que nao depende de contar dispositivos nem de adivinhar
        /// quem e quem.
        /// </summary>
        private static async Task AcrescentarSomenteXInputAsync(
            List<ControllerInfo> encontrados, HashSet<string> containersHid)
        {
            List<int> slots;
            try { slots = Native.SlotsXInputConectados(); }
            catch { return; }
            if (slots.Count == 0) return;

            var xusb = await DispositivosXusbAsync();

            List<(string Nome, string Instancia, string Container)> somente;
            if (xusb.Count == 0)
            {
                // Sem lista de XUSB nao ha container para comparar. Ainda assim, se o
                // XInput ve controle e o HID nao viu nenhum, nao existe com o que
                // confundir e os slots sao todos de controles invisiveis ao HID.
                if (encontrados.Count > 0) return;
                somente = slots.Select(_ => ((string)null, (string)null, (string)null)).ToList();
            }
            else
            {
                somente = xusb
                    .Where(x => string.IsNullOrEmpty(x.Container) || !containersHid.Contains(x.Container))
                    .Select(x => (x.Nome, x.Instancia, x.Container))
                    .ToList();
            }
            if (somente.Count == 0 || somente.Count > slots.Count) return;

            // na pratica os slots ocupados por controle HID vem primeiro na contagem do
            // XInput, entao o que sobra na ponta pertence a estes
            var meus = slots.Skip(slots.Count - somente.Count).ToList();
            var reserva = NomesDisponiveis(encontrados);

            for (int i = 0; i < somente.Count; i++)
            {
                encontrados.Add(new ControllerInfo
                {
                    Address = 0,
                    XInputSlot = meus[i],
                    InstanceId = somente[i].Instancia,
                    ContainerId = somente[i].Container,
                    Name = EscolherNome(somente[i].Nome, reserva, meus[i]),
                    LastSeen = DateTime.Now
                });
            }
        }

        /// <summary>Controles publicados pela interface do XUSB, com container e id.</summary>
        private static async Task<List<(string Container, string Nome, string Instancia)>>
            DispositivosXusbAsync()
        {
            var lista = new List<(string, string, string)>();
            try
            {
                var props = new[] { PropContainer, PropInstancia };
                var seletor = "System.Devices.InterfaceClassGuid:=\"" + GuidXusb + "\""
                            + " AND System.Devices.InterfaceEnabled:=System.StructuredQueryType.Boolean#True";
                var achados = await DeviceInformation.FindAllAsync(seletor, props);
                foreach (var d in achados)
                {
                    d.Properties.TryGetValue(PropContainer, out var container);
                    d.Properties.TryGetValue(PropInstancia, out var instancia);
                    lista.Add((Texto(container), d.Name, instancia as string));
                }
            }
            catch { }
            return lista;
        }

        /// <summary>O container vem como Guid, nao como texto, dependendo da consulta.</summary>
        private static string Texto(object valor) => valor switch
        {
            null => string.Empty,
            Guid g => g.ToString("B"),
            _ => valor.ToString()
        };

        /// <summary>
        /// O XInput nao informa nome nenhum, e o nome da interface do XUSB costuma ser
        /// generico. O Gaming.Input sabe o nome comercial, mas lista tambem os controles
        /// ja achados pelo HID -- por isso os repetidos saem antes.
        /// </summary>
        private static List<string> NomesDisponiveis(List<ControllerInfo> jaEncontrados)
        {
            var nomes = new List<string>();
            try
            {
                var usados = new HashSet<string>(
                    jaEncontrados.Select(c => c.Name ?? string.Empty), StringComparer.OrdinalIgnoreCase);
                foreach (var raw in Windows.Gaming.Input.RawGameController.RawGameControllers)
                {
                    var nome = raw.DisplayName;
                    if (string.IsNullOrWhiteSpace(nome) || !usados.Add(nome)) continue;
                    nomes.Add(nome);
                }
            }
            catch { }
            return nomes;
        }

        private static string EscolherNome(string doXusb, List<string> reserva, int slot)
        {
            if (reserva.Count > 0)
            {
                var nome = reserva[0];
                reserva.RemoveAt(0);
                return nome;
            }
            if (!string.IsNullOrWhiteSpace(doXusb)) return doXusb;
            return $"Controle {slot + 1}";
        }

        private readonly record struct GamepadHid(
            string IdHid, string IdInstancia, string Nome, ulong Endereco, string Container);

        /// <summary>Todo dispositivo HID que se declara controle, com ou sem Bluetooth.</summary>
        private static async Task<List<GamepadHid>> GamepadsHidAsync()
        {
            var lista = new List<GamepadHid>();
            var vistos = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            ushort[] usages = { UsageGamepad, UsageJoystick, UsageMultiAxis };
            var props = new[] { PropInstancia, PropContainer };

            foreach (var usage in usages)
            {
                IReadOnlyList<DeviceInformation> achados;
                try
                {
                    var seletor = HidDevice.GetDeviceSelector(UsagePageGenericDesktop, usage);
                    achados = await DeviceInformation.FindAllAsync(seletor, props);
                }
                catch { continue; }

                foreach (var d in achados)
                {
                    if (!vistos.Add(d.Id)) continue;
                    d.Properties.TryGetValue(PropInstancia, out var instancia);
                    d.Properties.TryGetValue(PropContainer, out var container);
                    lista.Add(new GamepadHid(d.Id, instancia as string, d.Name,
                        ExtractAddress(d.Id), Texto(container)));
                }
            }
            return lista;
        }

        private static async Task<HashSet<ulong>> GamepadAddressesAsync()
        {
            var set = new HashSet<ulong>();
            ushort[] usages = { UsageGamepad, UsageJoystick, UsageMultiAxis };

            foreach (var usage in usages)
            {
                string selector;
                try { selector = HidDevice.GetDeviceSelector(UsagePageGenericDesktop, usage); }
                catch { continue; }

                IReadOnlyList<DeviceInformation> found;
                try { found = await DeviceInformation.FindAllAsync(selector); }
                catch { continue; }

                foreach (var d in found)
                {
                    ulong a = ExtractAddress(d.Id);
                    if (a != 0) set.Add(a);
                }
            }
            return set;
        }

        internal static ulong ExtractAddress(string interfaceId)
        {
            if (string.IsNullOrEmpty(interfaceId)) return 0;
            foreach (Match m in AddressPattern.Matches(interfaceId))
            {
                if (!ulong.TryParse(m.Groups[1].Value,
                        System.Globalization.NumberStyles.HexNumber,
                        System.Globalization.CultureInfo.InvariantCulture, out var v))
                    continue;
                if (v != 0) return v;
            }
            return 0;
        }
    }

    /// <summary>
    /// Lembra os controles ja vistos para que continuem na lista mesmo desligados,
    /// quando nao existe nenhum HID para descobrir.
    /// </summary>
    public sealed class KnownControllers
    {
        private static readonly string Dir = AppPaths.DataDir;
        private static readonly string FilePath = AppPaths.File("controllers.json");

        private readonly List<ControllerInfo> _items = new();

        public IReadOnlyList<ControllerInfo> Items => _items;

        public void Load()
        {
            try
            {
                if (!File.Exists(FilePath)) return;
                var loaded = JsonSerializer.Deserialize<List<ControllerInfo>>(File.ReadAllText(FilePath));
                if (loaded != null) _items.AddRange(loaded.Where(i => i != null));
            }
            catch { }
        }

        /// <summary>Funde a descoberta com o que ja era conhecido. True se algo mudou.</summary>
        public bool Merge(IEnumerable<ControllerInfo> discovered)
        {
            bool dirty = false;
            foreach (var d in discovered)
            {
                var existing = _items.FirstOrDefault(i => i.Key == d.Key);
                if (existing == null)
                {
                    _items.Add(d);
                    dirty = true;
                }
                else
                {
                    if (existing.Name != d.Name) { existing.Name = d.Name; dirty = true; }
                    existing.HidId = d.HidId;
                    existing.InstanceId = d.InstanceId;
                    existing.XInputSlot = d.XInputSlot;
                    existing.ContainerId = d.ContainerId;
                    existing.LastSeen = d.LastSeen;
                }
            }
            if (dirty) Save();
            return dirty;
        }

        public void Touch(ulong address)
        {
            var it = _items.FirstOrDefault(i => i.Address == address);
            if (it != null) it.LastSeen = DateTime.Now;
        }

        public void Save()
        {
            try
            {
                Directory.CreateDirectory(Dir);
                File.WriteAllText(FilePath, JsonSerializer.Serialize(_items));
            }
            catch { }
        }
    }
}

File: src-tauri/src/device/discovery.rs
//! Quem esta ligado agora, seja qual for a marca ou a forma de ligacao.
//!
//! A descoberta nao reconhece modelo: pega o que se declara controle. Dispositivo HID
//! com uso de gamepad, joystick ou multi-eixo entra pela primeira via; quem so existe
//! para o XInput entra pela segunda. O nome exibido vem do proprio sistema.

use windows::core::{Result, HSTRING};
use windows::Devices::Enumeration::DeviceInformation;
use windows::Devices::HumanInterfaceDevice::HidDevice;
use windows_collections::IIterable;

use super::{hid, pnp, xinput};

/// Interface que o Windows publica para todo controle atendido pelo XUSB.
pub(crate) const GUID_XUSB: &str = "{EC87F1E3-C13B-4100-B5F7-8B84D54260CB}";

#[derive(Debug, Clone, Default)]
pub struct Controle {
    /// Endereco Bluetooth. Zero quando o controle so aparece por cabo ou dongle.
    pub endereco: u64,
    pub nome: String,
    pub id_hid: String,
    pub id_instancia: String,
    pub container: String,
    /// Slot do XInput, quando o controle so existe por ali. -1 para os demais.
    pub slot_xinput: i32,
}

impl Controle {
    /// Identidade estavel.
    ///
    /// Com Bluetooth e o endereco; sem ele, o proprio id da interface, que ja carrega
    /// fabricante, produto e a parte especifica daquela conexao.
    pub fn chave(&self) -> String {
        if self.endereco != 0 {
            return format!("{:012x}", self.endereco);
        }
        if self.slot_xinput >= 0 {
            return format!("xinput:{}", self.slot_xinput);
        }
        format!("hid:{}", chave_do_hid(&self.id_hid))
    }

    pub fn endereco_bonito(&self) -> Option<String> {
        if self.endereco == 0 {
            return None;
        }
        let b = self.endereco;
        Some(
            (0..6)
                .map(|i| format!("{:02X}", (b >> ((5 - i) * 8)) & 0xFF))
                .collect::<Vec<_>>()
                .join(":"),
        )
    }
}

fn chave_do_hid(id: &str) -> String {
    if id.is_empty() {
        return "desconhecido".to_string();
    }
    let limpo: String = id
        .chars()
        .filter(|c| c.is_alphanumeric())
        .map(|c| c.to_ascii_lowercase())
        .collect();
    if limpo.len() <= 40 {
        limpo
    } else {
        limpo[limpo.len() - 40..].to_string()
    }
}

/// Um bloco de doze hexadecimais delimitado por & ou _ dentro do id da interface.
pub fn extrair_endereco(id: &str) -> u64 {
    let bytes: Vec<char> = id.chars().collect();
    let delimitador = |c: char| c == '&' || c == '_';

    for inicio in 0..bytes.len() {
        if !delimitador(bytes[inicio]) {
            continue;
        }
        let fim = inicio + 13;
        if fim >= bytes.len() || !delimitador(bytes[fim]) {
            continue;
        }
        let trecho: String = bytes[inicio + 1..fim].iter().collect();
        if !trecho.chars().all(|c| c.is_ascii_hexdigit()) {
            continue;
        }
        if let Ok(v) = u64::from_str_radix(&trecho, 16) {
            if v != 0 {
                return v;
            }
        }
    }
    0
}

pub fn descobrir() -> Vec<Controle> {
    descobrir_com(&super::gatt::pareados())
}

/// A varredura, com a lista de pareados vinda de fora.
///
/// `pareados` so serve para dar nome: ela custa uma abertura de dispositivo Bluetooth
/// por aparelho e o nome de um controle nao muda entre uma varredura e a seguinte. Quem
/// chama guarda a lista e a renova quando ela envelhece -- e essa separacao e o que
/// permite varrer de segundos em segundos sem acordar o radio a toa.
pub fn descobrir_com(pareados: &[(u64, String)]) -> Vec<Controle> {
    // Sem HID ainda pode haver controle: o XUSB e um mundo a parte, conferido mais
    // abaixo. Desistir aqui era justamente o que escondia controle de dongle.
    let achados = gamepads_hid();

    let mut resultado: Vec<Controle> = Vec::new();
    let mut containers_hid: Vec<String> = Vec::new();

    for g in achados {
        let nome = pareados
            .iter()
            .find(|(endereco, _)| *endereco == g.endereco && g.endereco != 0)
            .map(|(_, nome)| nome.clone())
            .unwrap_or_else(|| {
                if g.nome.trim().is_empty() {
                    "Controle".to_string()
                } else {
                    g.nome.clone()
                }
            });

        containers_hid.push(g.container.clone());
        resultado.push(Controle {
            endereco: g.endereco,
            nome,
            id_hid: g.id_hid,
            id_instancia: g.id_instancia,
            container: g.container,
            slot_xinput: -1,
        });
    }

    acrescentar_somente_xinput(&mut resultado, &containers_hid);
    resultado
}

struct GamepadHid {
    id_hid: String,
    id_instancia: String,
    container: String,
    nome: String,
    endereco: u64,
}

fn gamepads_hid() -> Vec<GamepadHid> {
    let usos = [hid::USO_GAMEPAD, hid::USO_JOYSTICK, hid::USO_MULTI_EIXO];
    let mut lista = Vec::new();
    let mut vistos: Vec<String> = Vec::new();

    for uso in usos {
        let Ok(encontrados) = por_uso(uso) else { continue };
        for g in encontrados {
            if vistos.iter().any(|v| v.eq_ignore_ascii_case(&g.id_hid)) {
                continue;
            }
            vistos.push(g.id_hid.clone());
            lista.push(g);
        }
    }
    lista
}

fn por_uso(uso: u16) -> Result<Vec<GamepadHid>> {
    let seletor = HidDevice::GetDeviceSelector(hid::PAGINA_DESKTOP_GENERICO, uso)?;
    let chaves = [pnp::CHAVE_INSTANCIA, pnp::CHAVE_CONTAINER];
    let props: Vec<HSTRING> = chaves.iter().map(|c| HSTRING::from(*c)).collect();
    let achados = DeviceInformation::FindAllAsyncAqsFilterAndAdditionalProperties(
        &seletor,
        &IIterable::<HSTRING>::from(props),
    )?
    .join()?;

    let mut saida = Vec::new();
    for info in achados {
        let id = info.Id()?.to_string();
        let propriedades = info.Properties()?;
        let texto = |chave: &str| -> String {
            let chave = HSTRING::from(chave);
            propriedades
                .HasKey(&chave)
                .ok()
                .and_then(|tem| tem.then(|| propriedades.Lookup(&chave).ok()).flatten())
                .and_then(|v| {
                    use windows::core::Interface;
                    let pv: windows::Foundation::IPropertyValue = v.cast().ok()?;
                    match pv.Type().ok()? {
                        windows::Foundation::PropertyType::String => {
                            pv.GetString().ok().map(|s| s.to_string())
                        }
                        windows::Foundation::PropertyType::Guid => {
                            pv.GetGuid().ok().map(|g| format!("{g:?}"))
                        }
                        _ => None,
                    }
                })
                .unwrap_or_default()
        };

        saida.push(GamepadHid {
            endereco: extrair_endereco(&id),
            id_instancia: texto(pnp::CHAVE_INSTANCIA),
            container: texto(pnp::CHAVE_CONTAINER),
            nome: info.Name()?.to_string(),
            id_hid: id,
        });
    }
    Ok(saida)
}

/// Acrescenta controles que existem apenas para o XInput.
///
/// Quem usa o driver do Xbox 360 -- o caso dos dongles que emulam esse controle -- nao
/// e dispositivo HID. A busca acima passa direto por ele, e sem isto esse controle
/// simplesmente nao existe para o app.
///
/// A comparacao e por container, que representa o aparelho fisico: um controle atendido
/// por HID e por XUSB publica as duas interfaces sob o mesmo container. Container de
/// XUSB que nao apareceu no HID e, por definicao, controle que so o XInput enxerga --
/// criterio que nao depende de contar dispositivos nem de adivinhar quem e quem.
fn acrescentar_somente_xinput(encontrados: &mut Vec<Controle>, containers_hid: &[String]) {
    let slots = xinput::slots_conectados();
    if slots.is_empty() {
        return;
    }

    let xusb = dispositivos_xusb();

    let somente: Vec<(String, String, String)> = if xusb.is_empty() {
        // Sem lista de XUSB nao ha container para comparar. Ainda assim, se o XInput ve
        // controle e o HID nao viu nenhum, nao existe com o que confundir.
        if !encontrados.is_empty() {
            return;
        }
        slots.iter().map(|_| (String::new(), String::new(), String::new())).collect()
    } else {
        xusb.into_iter()
            .filter(|(container, _, _)| {
                container.is_empty()
                    || !containers_hid.iter().any(|c| c.eq_ignore_ascii_case(container))
            })
            .collect()
    };

    if somente.is_empty() || somente.len() > slots.len() {
        return;
    }

    // na pratica os slots ocupados por controle HID vem primeiro na contagem do XInput,
    // entao o que sobra na ponta pertence a estes
    let meus = &slots[slots.len() - somente.len()..];
    let mut reserva = nomes_disponiveis(encontrados);

    for (i, (container, nome_xusb, instancia)) in somente.into_iter().enumerate() {
        let slot = meus[i];
        encontrados.push(Controle {
            endereco: 0,
            nome: escolher_nome(&nome_xusb, &mut reserva, slot),
            id_hid: String::new(),
            id_instancia: instancia,
            container,
            slot_xinput: slot as i32,
        });
    }
}

/// Devolve (container, nome, instancia) de cada interface do XUSB ativa.
fn dispositivos_xusb() -> Vec<(String, String, String)> {
    (|| -> Result<Vec<(String, String, String)>> {
        let seletor = HSTRING::from(format!(
            "System.Devices.InterfaceClassGuid:=\"{GUID_XUSB}\" AND \
             System.Devices.InterfaceEnabled:=System.StructuredQueryType.Boolean#True"
        ));
        let chaves = [pnp::CHAVE_CONTAINER, pnp::CHAVE_INSTANCIA];
        let props: Vec<HSTRING> = chaves.iter().map(|c| HSTRING::from(*c)).collect();
        let achados = DeviceInformation::FindAllAsyncAqsFilterAndAdditionalProperties(
            &seletor,
            &IIterable::<HSTRING>::from(props),
        )?
        .join()?;

        let mut saida = Vec::new();
        for info in achados {
            let propriedades = info.Properties()?;
            let texto = |chave: &str| -> String {
                use windows::core::Interface;
                let chave = HSTRING::from(chave);
                propriedades
                    .HasKey(&chave)
                    .ok()
                    .and_then(|tem| tem.then(|| propriedades.Lookup(&chave).ok()).flatten())
                    .and_then(|v| {
                        let pv: windows::Foundation::IPropertyValue = v.cast().ok()?;
                        match pv.Type().ok()? {
                            windows::Foundation::PropertyType::String => {
                                pv.GetString().ok().map(|s| s.to_string())
                            }
                            windows::Foundation::PropertyType::Guid => {
                                pv.GetGuid().ok().map(|g| format!("{g:?}"))
                            }
                            _ => None,
                        }
                    })
                    .unwrap_or_default()
            };
            saida.push((
                texto(pnp::CHAVE_CONTAINER),
                info.Name()?.to_string(),
                texto(pnp::CHAVE_INSTANCIA),
            ));
        }
        Ok(saida)
    })()
    .unwrap_or_default()
}

/// O XInput nao informa nome nenhum, e o nome da interface do XUSB costuma ser generico.
fn nomes_disponiveis(ja_encontrados: &[Controle]) -> Vec<String> {
    let usados: Vec<String> = ja_encontrados.iter().map(|c| c.nome.to_lowercase()).collect();
    pnp::nos_com_bateria()
        .into_iter()
        .map(|no| no.nome)
        .filter(|nome| !nome.trim().is_empty() && !usados.contains(&nome.to_lowercase()))
        .collect()
}

fn escolher_nome(do_xusb: &str, reserva: &mut Vec<String>, slot: u32) -> String {
    if !reserva.is_empty() {
        return reserva.remove(0);
    }
    if !do_xusb.trim().is_empty() {
        return do_xusb.to_string();
    }
    format!("Controle {}", slot + 1)
}
